File: capture-page/js/crypto.js
// SPDX-FileCopyrightText: 2026 Jasper Curry
//
// SPDX-License-Identifier: Apache-2.0

// Page-side end-to-end encryption for the capture relay (build steps 3 + 5).
//
// The content_key (AES-256-GCM, 32 bytes) is minted by the Pi and delivered to
// this page ONLY in the URL fragment — the one part of a URL browsers never
// transmit — so the relay never sees it and stores ciphertext only. This module
// encrypts the recorded WAV here and computes a PLAINTEXT integrity tag (length
// + SHA-256) that the Pi verifies AFTER decrypt, BEFORE analysis. The matching
// Pi-side decrypt + cross-language round-trip proof land in step 5
// (jasper/capture_relay/crypto.py); the wire format is the contract between them:
//
//   key   : 32 raw bytes, base64url in the fragment
//   blob  : IV(12 bytes) ‖ AES-256-GCM(ciphertext ‖ 16-byte tag)
//   hash  : lowercase hex SHA-256 over the PLAINTEXT WAV bytes
//   length: plaintext WAV byte count
//
// Pinned by tests/js/capture_crypto_test.mjs (and the cross-language test in
// step 5).

const IV_BYTES = 12;

export function base64urlToBytes(b64url) {
  const b64 = String(b64url).replace(/-/g, "+").replace(/_/g, "/");
  const pad = b64.length % 4 === 0 ? "" : "=".repeat(4 - (b64.length % 4));
  const bin = atob(b64 + pad);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

export function bytesToHex(bytes) {
  const u8 = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);
  let s = "";
  for (const b of u8) s += b.toString(16).padStart(2, "0");
  return s;
}

export async function sha256Hex(bytes) {
  const u8 = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);
  const digest = await crypto.subtle.digest("SHA-256", u8);
  return bytesToHex(new Uint8Array(digest));
}

export async function importContentKey(b64urlKey) {
  const raw = base64urlToBytes(b64urlKey);
  if (raw.length !== 32) {
    throw new Error(`content_key must be 32 bytes, got ${raw.length}`);
  }
  return crypto.subtle.importKey("raw", raw, { name: "AES-GCM" }, false, ["encrypt"]);
}

// Encrypt plaintext WAV bytes. Returns the relay blob (IV ‖ ciphertext) and the
// plaintext integrity the Pi verifies after decrypt.
export async function encryptWav(key, wavBytes) {
  const plaintext = wavBytes instanceof Uint8Array ? wavBytes : new Uint8Array(wavBytes);
  const iv = crypto.getRandomValues(new Uint8Array(IV_BYTES));
  const ctBuf = await crypto.subtle.encrypt({ name: "AES-GCM", iv }, key, plaintext);
  const ciphertext = new Uint8Array(ctBuf);
  const blob = new Uint8Array(iv.length + ciphertext.length);
  blob.set(iv, 0);
  blob.set(ciphertext, iv.length);
  return {
    blob,
    plaintextLen: plaintext.length,
    sha256: await sha256Hex(plaintext),
  };
}

export const _format = { IV_BYTES };

File: jasper/capture_relay/crypto.py
# SPDX-FileCopyrightText: 2026 Jasper Curry
#
# SPDX-License-Identifier: Apache-2.0

"""Pi-side end-to-end decryption + integrity for the capture relay (step 5).

The phone encrypts the recorded WAV under a `content_key` the **Pi** minted and
delivered only in the page's URL fragment — so the relay stores ciphertext only
and never sees the key (plan §8). This module is the Pi's matching half: it mints
the key, decrypts the relay-pulled blob, and verifies the **plaintext** integrity
the phone attached, **before** the WAV reaches analysis. An intact-but-corrupt or
truncated capture fails loud here (plan §9, §15), never a silently-wrong number.

Wire format — the contract with the page (capture-page/js/crypto.js):

  key   : 32 raw bytes, base64url in the fragment
  blob  : IV(12 bytes) ‖ AES-256-GCM(ciphertext ‖ 16-byte tag)
  hash  : lowercase hex SHA-256 over the PLAINTEXT WAV bytes
  length: plaintext WAV byte count

AES-GCM already authenticates the ciphertext (a tampered blob fails to decrypt);
the explicit plaintext length + SHA-256 is the second, plan-mandated check that
also catches a valid-but-wrong-length payload and is the "decrypted WAV is
bit-identical" proof (pinned cross-language by tests/test_capture_relay_crypto.py
against the page's WebCrypto encryptor).
"""
from __future__ import annotations

import base64
import hashlib
import hmac
import secrets

from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

CONTENT_KEY_BYTES = 32  # AES-256
IV_BYTES = 12
GCM_TAG_BYTES = 16


class DecryptError(ValueError):
    """The relay blob could not be decrypted (wrong key or tampered ciphertext)."""


class IntegrityError(ValueError):
    """Decryption succeeded but the plaintext failed its length/SHA-256 check."""


# --- key lifecycle ------------------------------------------------------------


def generate_content_key() -> bytes:
    """Mint a fresh AES-256-GCM content key with a CSPRNG (plan §11)."""
    return secrets.token_bytes(CONTENT_KEY_BYTES)


def content_key_to_b64url(key: bytes) -> str:
    """Encode the key for the URL fragment (unpadded base64url)."""
    if len(key) != CONTENT_KEY_BYTES:
        raise ValueError(f"content_key must be {CONTENT_KEY_BYTES} bytes")
    return base64.urlsafe_b64encode(key).rstrip(b"=").decode("ascii")


def content_key_from_b64url(value: str) -> bytes:
    """Decode a base64url content key (padding-tolerant)."""
    padded = value + "=" * (-len(value) % 4)
    raw = base64.urlsafe_b64decode(padded)
    if len(raw) != CONTENT_KEY_BYTES:
        raise ValueError(f"content_key must decode to {CONTENT_KEY_BYTES} bytes")
    return raw


# --- decrypt + verify ---------------------------------------------------------


def decrypt_blob(content_key: bytes, blob: bytes) -> bytes:
    """Decrypt an ``IV ‖ ciphertext`` relay blob. Raises DecryptError on failure."""
    if len(content_key) != CONTENT_KEY_BYTES:
        raise DecryptError(f"content_key must be {CONTENT_KEY_BYTES} bytes")
    if len(blob) < IV_BYTES + GCM_TAG_BYTES:
        raise DecryptError("blob too short to contain IV + GCM tag")
    iv = blob[:IV_BYTES]
    ciphertext = blob[IV_BYTES:]
    try:
        return AESGCM(content_key).decrypt(iv, ciphertext, None)
    except InvalidTag as exc:
        raise DecryptError(
            "AES-GCM authentication failed (wrong key or tampered ciphertext)"
        ) from exc


def verify_integrity(
    plaintext: bytes, expected_len: int, expected_sha256_hex: str
) -> None:
    """Verify the plaintext length + SHA-256 the phone attached. Loud on mismatch."""
    if len(plaintext) != expected_len:
        raise IntegrityError(
            f"plaintext length {len(plaintext)} != expected {expected_len}"
        )
    actual = hashlib.sha256(plaintext).hexdigest()
    if not hmac.compare_digest(actual, (expected_sha256_hex or "").lower()):
        raise IntegrityError("plaintext SHA-256 mismatch")


def decrypt_and_verify(
    content_key: bytes,
    blob: bytes,
    expected_len: int,
    expected_sha256_hex: str,
) -> bytes:
    """Decrypt then verify integrity, returning the WAV bytes. The single call
    the Pi pull path uses before handing the WAV to analysis."""
    plaintext = decrypt_blob(content_key, blob)
    verify_integrity(plaintext, expected_len, expected_sha256_hex)
    return plaintext
